service/history/replication/raw_task_converter.go

```diff
@@ -712,3 +637,323 @@ func convertGetHistoryError(
 		return err
 	}
 }
+
+func newSyncVersionedTransitionTaskConverter(
+	ctx context.Context,
+	shardContext shard.Context,
+	historyEngine shard.Engine,
+	taskInfo *tasks.SyncVersionedTransitionTask,
+	shardID int32,
+	workflowCache wcache.Cache,
+	eventBlobCache persistence.XDCCache,
+	replicationCache ProgressCache,
+	targetClusterID int32,
+	executionManager persistence.ExecutionManager,
+	mutableState workflow.MutableState,
+	logger log.Logger,
+) *syncVersionedTransitionTaskConverter {
+	progress := replicationCache.Get(taskInfo.RunID, targetClusterID)
+	return &syncVersionedTransitionTaskConverter{
+		ctx:              ctx,
+		shardContext:     shardContext,
+		historyEngine:    historyEngine,
+		taskInfo:         taskInfo,
+		shardID:          shardID,
+		workflowCache:    workflowCache,
+		eventBlobCache:   eventBlobCache,
+		replicationCache: replicationCache,
+		targetClusterID:  targetClusterID,
+		executionManager: executionManager,
+		mutableState:     mutableState,
+		logger:           logger,
+		progress:         progress,
+	}
+}
+
+func (c *syncVersionedTransitionTaskConverter) convert() (*replicationspb.ReplicationTask, error) {
+	// 1. task versioned transition not on current transition history
+	//   1.1 task has no events and no new run                   : no-op
+	//   1.2 task.VersionedTransition is found in cache 				 : verify [task]
+	//   1.3 task.VersionedTransition is not found in cache      : backfill events [task]
+	// 2. task versioned transition on current transition history
+	//   2.1 never synced                                        : SyncStateMutation: events [task, current]
+	//   2.2 task.VersionedTransition is found in cache          : verify [task]
+	//   2.3 last sync on current transition history             : SyncStateMutation: events (LCA(cache,current), current]
+	//   2.4 last sync not on current transition history         : SyncStateSnapshot: events (LCA(lastSync,current),current]
+
+	transitionHistory := c.mutableState.GetExecutionInfo().TransitionHistory
+	if !c.onCurrentBranch(c.taskInfo.VersionedTransition) {
+		// 1.1 task has no events and no new run
+		if c.taskInfo.FirstEventID == c.taskInfo.NextEventID && len(c.taskInfo.NewRunID) == 0 {
+			return nil, nil
+		}
+		if c.progress.TaskSent(c.taskInfo.VersionedTransition) {
+			return c.generateVerifyVersionedTransitionTask()
+		}
+		return c.generateBackfillHistoryTask()
+	}
+
+	lastSyncedTransition := c.progress.LastSyncedTransition()
+	if lastSyncedTransition == nil {
+		return c.generateSyncStateMutationFromTask(transitionHistory)
+	}
+	if c.progress.TaskSent(c.taskInfo.VersionedTransition) {
+		return c.generateVerifyVersionedTransitionTask()
+	}
+
+	lca, err := c.findLCAVersionHistoryItemFromCache()
+	if err != nil {
+		return nil, err
+	}
+	if c.onCurrentBranch(lastSyncedTransition) {
```

Comment:
we also need to check the the versioned Transition task we are syncing here is covered by the tombstone list in mutable state. Basically the task versioned transition >= the first versioned transition in executionInfo.subStateMachineTombstoneBatches. If not, we have to sync snapshot as well.

Reply:
will solve this in another PR combining with the SyncState API.